source/linker.cpp

```diff
@@ -0,0 +1,1098 @@
+    }
+
+    position.index = i += insn_word_count;
+  }
+
+  return SPV_SUCCESS;
+}
+
+static const char* AddressingModelStr(spv::AddressingModel model) {
+  switch (model) {
+  case spv::AddressingModel::Logical:    return "Logical";
+  case spv::AddressingModel::Physical32: return "Physical32";
+  case spv::AddressingModel::Physical64: return "Physical64";
+  case spv::AddressingModel::Max:        return "Unsupported";
+  }
+}
+
+static const char* MemoryModelStr(spv::MemoryModel model) {
+  switch (model) {
+  case spv::MemoryModel::Simple:  return "Simple";
+  case spv::MemoryModel::GLSL450: return "GLSL450";
+  case spv::MemoryModel::OpenCL:  return "OpenCL";
+  case spv::MemoryModel::Max:     return "Unsupported";
+  }
+}
+
+static bool IsCapabilityOpcode(spv::Op opcode) {
+  return opcode == spv::Op::OpCapability;
+}
+
+static bool IsExtensionOpcode(spv::Op opcode) {
+  return opcode == spv::Op::OpExtension;
+}
+
+static bool IsExtInstImportOpcode(spv::Op opcode) {
+  return opcode == spv::Op::OpExtInstImport;
+}
+
+static bool IsEntryPointOpcode(spv::Op opcode) {
+  return opcode == spv::Op::OpEntryPoint;
+}
+
+static bool IsExecutionModeOpcode(spv::Op opcode) {
+  return opcode == spv::Op::OpExecutionMode;
+}
+
+static bool IsDebugAOpcode(spv::Op opcode) {
+  switch (opcode) {
+  case spv::Op::OpString:
+  case spv::Op::OpSourceExtension:
+  case spv::Op::OpSource:
+  case spv::Op::OpSourceContinued:
+    return true;
+  default:
+    return false;
+  }
+}
+
+static bool IsDebugBOpcode(spv::Op opcode) {
+  return opcode == spv::Op::OpName || opcode == spv::Op::OpMemberName;
+}
+
+static bool IsTypeOrConstantOpcode(spv::Op opcode) {
+  switch (opcode) {
+  case spv::Op::OpLine:
+  case spv::Op::OpTypeVoid:
+  case spv::Op::OpTypeBool:
+  case spv::Op::OpTypeInt:
+  case spv::Op::OpTypeFloat:
+  case spv::Op::OpTypeVector:
+  case spv::Op::OpTypeMatrix:
+  case spv::Op::OpTypeImage:
+  case spv::Op::OpTypeSampler:
+  case spv::Op::OpTypeSampledImage:
+  case spv::Op::OpTypeArray:
+  case spv::Op::OpTypeRuntimeArray:
+  case spv::Op::OpTypeStruct:
+  case spv::Op::OpTypeOpaque:
+  case spv::Op::OpTypePointer:
+  case spv::Op::OpTypeFunction:
+  case spv::Op::OpTypeEvent:
+  case spv::Op::OpTypeDeviceEvent:
+  case spv::Op::OpTypeReserveId:
+  case spv::Op::OpTypeQueue:
+  case spv::Op::OpTypePipe:
+  case spv::Op::OpTypeForwardPointer:
+  case spv::Op::OpConstantTrue:
+  case spv::Op::OpConstantFalse:
+  case spv::Op::OpConstant:
+  case spv::Op::OpConstantComposite:
+  case spv::Op::OpConstantSampler:
+  case spv::Op::OpConstantNull:
+  case spv::Op::OpSpecConstantTrue:
+  case spv::Op::OpSpecConstantFalse:
+  case spv::Op::OpSpecConstant:
+  case spv::Op::OpSpecConstantComposite:
+  case spv::Op::OpSpecConstantOp:
+  case spv::Op::OpVariable:
+  case spv::Op::OpUndef:
+    return true;
+  default:
+    return false;
+  }
+}
+
+static spv_result_t MergeModules(spv_context_t& context,
```

Comment:
There's a lot of custom code here to sort out instructions into their own sections.  I think you could shorten and simplify a lot of this code by reusing the Module class from the optimizer.  See source/opt/module.h.

Its internal representation already segregate the instructions in different logical layout sections.  You make a Module with BuildModule from opt/build_module.h.
There are already iterator ranges for things like debug instructions, types, annotations, entry_points.  I would encourage adding methods to get iterator ranges for the other layout sections.